web/handlers.go

```diff
@@ -177,15 +178,52 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		w.Header().Set("X-Frame-Options", "SAMEORIGIN")
 
 		// Add unsafe-eval to the content security policy for faster source maps in development mode
-		devCSP := ""
+		devCSPMap := make(map[string]bool)
```

Comment:
This whole logic is getting too big to be contained inside `ServeHTTP`. Let's refactor this outside to a separate function which takes the `*c.App.Config().ServiceSettings.DeveloperFlags` setting and returns the devCSP value. That would also make it easier for you to test this function.

Reply:
I take the fault for the complexity add-on here :)

Reply:
I've refactored this into a separate function and have added tests for the same. However, I'm running into the following error while using `c.Logger.Warn` :

`
--- FAIL: TestGenerateDevCSP (0.53s)
    --- FAIL: TestGenerateDevCSP/dev_mode (0.53s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x28e3035]

goroutine 8236 [running]:
testing.tRunner.func1.2({0x2b04bc0, 0x4525800})
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1212 +0x218
panic({0x2b04bc0, 0x4525800})
        /usr/local/Cellar/go/1.17.2/libexec/src/runtime/panic.go:1038 +0x215
github.com/mattermost/mattermost-server/v6/shared/mlog.(*Logger).Warn(...)
        /Users/spotdraft/personal_workspace/mattermost-server/shared/mlog/mlog.go:302
command-line-arguments.(*Handler).generateDevCSP(0x326e110, {{0x32b0410, 0xc00231c358}, 0xc0037b6400, 0x0, 0x0, 0x0, {0x0, 0x0}})
        /Users/spotdraft/personal_workspace/mattermost-server/web/handlers.go:107 +0x7f5
command-line-arguments.TestGenerateDevCSP.func1(0x0)
        /Users/spotdraft/personal_workspace/mattermost-server/web/handlers_test.go:412 +0x1b5
testing.tRunner(0xc000fc76c0, 0x2eca960)
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1306 +0x35a
FAIL    command-line-arguments  3.914s
FAIL

`

I'm not able to figure out what the issue is here. Can one of you please help me out here?

Reply:
The error says: "invalid memory address or nil pointer dereference". So that means it is trying to dereference a nil pointer. If you look at the code, you can see it's coming from the line `c.Logger.Warn`. And you can see that Logger is not set for the context when you are initializing the test.

Also, there's no need to make it a method of `*Handler`, just a normal function is fine. We can just pass the context, no need to dereference that, and get the string.